quickwit-indexing/src/source/kinesis/shard_consumer.rs

```diff
@@ -0,0 +1,451 @@
+    /// The consumer has reached the latest record in the shard and stops if `eof_enabled` is set
+    /// to true.
+    ShardEOF(String),
+}
+
+#[derive(Default)]
+struct ShardConsumerState {
+    /// The sequence number of the last record transferred.
+    current_sequence_number: Option<String>,
+    /// The number of milliseconds the last `GetRecords` response is from the tip of the stream.
+    lag_millis: Option<i64>,
+    /// Number of bytes transferred by the consumer.
+    num_bytes_transferred: u64,
+    /// Number of records transferred by the consumer.
+    num_records_transferred: u64,
+    /// The shard iterator value that will be used for the next call to `GetRecords`.
+    next_shard_iterator: Option<String>,
+}
+
+struct ShardConsumer {
+    stream_name: String,
+    shard_id: String,
+    /// Sequence number of the last record processed. Consumption of the shard is resumed right
+    /// after this sequence number.
+    from_sequence_number_exclusive: Option<String>,
+    /// When this value is set to true, the consumer stops after reaching the last (most recent)
+    /// record in the shard.
+    eof_enabled: bool,
+    state: ShardConsumerState,
+    kinesis_client: Box<dyn Kinesis + Send + Sync>,
+    sink: Mailbox<ShardConsumerMessage>,
+}
+
+impl ShardConsumer {
+    fn new(
+        stream_name: String,
+        shard_id: String,
+        from_sequence_number_exclusive: Option<String>,
+        eof_enabled: bool,
+        kinesis_client: Box<dyn Kinesis + Send + Sync>,
+        sink: Mailbox<ShardConsumerMessage>,
+    ) -> Self {
+        Self {
+            stream_name,
+            shard_id,
+            from_sequence_number_exclusive,
+            state: Default::default(),
+            eof_enabled,
+            kinesis_client,
+            sink,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Loop {}
+
+impl Actor for ShardConsumer {
+    type Message = Loop;
+    type ObservableState = serde_json::Value;
+
+    fn name(&self) -> String {
+        "KinesisShardConsumer".to_string()
+    }
+
+    fn observable_state(&self) -> Self::ObservableState {
+        json!({
+            "stream_name": self.stream_name,
+            "shard_id": self.shard_id,
+            "current_sequence_number": self.state.current_sequence_number,
+            "lag_millis": self.state.lag_millis,
+            "num_bytes_transferred": self.state.num_bytes_transferred,
+            "num_records_transferred": self.state.num_records_transferred,
+        })
+    }
+}
+
+#[async_trait]
+impl AsyncActor for ShardConsumer {
+    async fn initialize(
+        &mut self,
+        ctx: &ActorContext<Self::Message>,
+    ) -> Result<(), ActorExitStatus> {
+        self.state.next_shard_iterator = get_shard_iterator(
+            &*self.kinesis_client,
+            &self.stream_name,
+            &self.shard_id,
+            self.from_sequence_number_exclusive.clone(),
+        )
+        .await?;
+        self.process_message(Loop {}, ctx).await?;
```

Comment:
Super nitpick. Probably just a taste story.

`struct Loop` and `Loop` works too.